lua/barbar/render.lua

```diff
@@ -702,91 +553,141 @@ local function generate_tabline(bufnrs, refocus)
   -- Create actual tabline string
   local result = ''
 
-  -- Add offset filler & text (for filetree/sidebar plugins)
+  -- Left offset
   if state.offset.left.width > 0 then
-    --- @type barbar.render.group
-    local offset = {hl = hl_tabline(state.offset.left.hl or 'BufferOffset'), text = ' ' .. state.offset.left.text}
-    local offset_available_width = state.offset.left.width - 2
-
-    result = result ..
-      groups_to_string(slice_groups_right({offset}, offset_available_width)) ..
-      (' '):rep(offset_available_width - strwidth(state.offset.left.text) + 1)
-  end
+    local hl = wrap_hl(state.offset.left.hl)
+    local offset_groups = { { hl = hl, text = (' '):rep(state.offset.left.width) } }
 
-  --- The highlight of the buffer tabpage fill
-  local hl_buffer_tabpage_fill = hl_tabline('BufferTabpageFill')
+    local content = { { hl = hl, text = state.offset.left.text } }
+    local content_max_width = state.offset.left.width - 2
 
-  --- @type barbar.render.group[]
-  local bufferline_groups = {{
-    hl = hl_buffer_tabpage_fill,
-    text = (' '):rep(layout.actual_width),
-  }}
+    offset_groups =
+      groups.insert_many(
+        offset_groups,
+        1,
+        groups.slice_right(content, content_max_width))
 
-  for _, group_clump in ipairs(group_clumps) do
-    bufferline_groups = groups_insert(bufferline_groups, group_clump.position, group_clump.groups)
+    result = result .. groups.to_string(offset_groups)
   end
 
-  do -- Crop to scroll region
-    local scroll_current = min(scroll.current, layout.scroll_max)
-    local buffers_end = layout.actual_width - scroll_current
+  -- Buffer tabs
+  do
+    --- @type barbar.render.group[]
+    local content = {
+      {
+        hl = HL.FILL,
+        text = (' '):rep(layout.buffers.width),
+      }
+    }
 
-    if buffers_end > layout.buffers_width then
-      bufferline_groups = slice_groups_right(bufferline_groups, scroll_current + layout.buffers_width)
+    do
+      local current_group_clump = nil
+      for _, group_clump in ipairs(group_clumps) do
+        -- We insert the current buffer after the others so it's always on top
```

Comment:
This is something I was trying to accomplish with the old `scroll_lock = {current = true}` in #408's first iteration. Good to see it implemented.